drivers/storage/digitalocean/storage/digitalocean_storage.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (d *driver) VolumeAttach(ctx types.Context, volumeID string, opts *types.VolumeAttachOpts) (*types.Volume, string, error) {
+	vol, _, err := d.client.Storage.GetVolume(volumeID)
+	if err != nil {
+		return nil, "", goof.WithError("error retrieving volume", err)
+	}
+
+	if len(vol.DropletIDs) > 0 {
+		if !opts.Force {
+			return nil, "", goof.New("volume already attached!")
+		}
+
+		_, _, err := d.client.StorageActions.Detach(volumeID)
+		if err != nil {
+			return nil, "", err
+		}
+	}
+
+	dropletID := mustInstanceIDID(ctx)
+	dropletIDI, err := strconv.Atoi(*dropletID)
+	if err != nil {
+		return nil, "", err
+	}
+
+	action, _, err := d.client.StorageActions.Attach(volumeID, dropletIDI)
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = d.waitForAction(volumeID, action)
+	if err != nil {
+		return nil, "", err
+	}
+
+	attachedVol, err := d.VolumeInspect(ctx, volumeID, &types.VolumeInspectOpts{
+		Attachments: types.VolAttReqTrue})
+	if err != nil {
+		return nil, "", goof.WithError("error retrieving volume", err)
+	}
+
+	return attachedVol, attachedVol.Attachments[0].DeviceName, nil
+}
+
+func (d *driver) VolumeDetach(ctx types.Context, volumeID string, opts *types.VolumeDetachOpts) (*types.Volume, error) {
+	vol, _, err := d.client.Storage.GetVolume(volumeID)
+	if err != nil {
+		return nil, goof.WithError("error getting volume", err)
+	}
+
+	if len(vol.DropletIDs) == 0 {
+		return nil, goof.WithError("volume already detached", err)
+	}
+
+	action, _, err := d.client.StorageActions.Detach(volumeID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.waitForAction(volumeID, action)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.Info("detached volume", volumeID)
+
+	detachedVol, err := d.VolumeInspect(ctx, volumeID, &types.VolumeInspectOpts{
+		Attachments: types.VolAttReqTrue,
+		Opts:        opts.Opts,
+	})
+	if err != nil {
+		return nil, goof.WithError("error getting volume information", err)
+	}
+
+	return detachedVol, nil
+}
+
+func (d *driver) Snapshots(ctx types.Context, opts types.Store) ([]*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotInspect(ctx types.Context, snapshotID string, opts types.Store) (*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotCopy(ctx types.Context, snapshotID string, snapshotName string, destinationID string, opts types.Store) (*types.Snapshot, error) {
+	return nil, types.ErrNotImplemented
+}
+
+func (d *driver) SnapshotRemove(ctx types.Context, snapshotID string, opts types.Store) error {
+	return types.ErrNotImplemented
+}
+
+func mustInstanceIDID(ctx types.Context) *string {
+	return &context.MustInstanceID(ctx).ID
+}
+
+func (d *driver) toTypesVolume(ctx types.Context, volume *godo.Volume, attachments types.VolumeAttachmentsTypes) *types.Volume {
+	// Collect attachment info for the volume
+	var atts []*types.VolumeAttachment
+	if attachments.Requested() {
+		for _, id := range volume.DropletIDs {
```

Comment:
Hi @protochron,

So the issue is that your tests are perhaps using an over-eager attachments mask, thus causing volumes to be filtered out by the framework and not returned to you. This is by design. See the [attachments related querying tests](https://github.com/codedellemc/libstorage/blob/master/drivers/storage/vfs/tests/vfs_test.go#L151-L377) for the VFS storage driver. I use the VFS driver as the default means of validating the framework since VFS is pure-software with no external dependencies.

What you need to do is review the bit mask constants for querying volumes with attachments. 

* [`VolumeAttachmentTypes` code definitions](https://github.com/codedellemc/libstorage/blob/master/api/types/types_drivers_storage.go#L11-L38)
* libStorage API reference for [querying volumes with attachment information](http://docs.libstorage.apiary.io/#reference/volumes/volumes-collection/get-with-attachments)

Finally, here is a series of [predefined masks](https://github.com/codedellemc/libstorage/blob/master/api/types/types_drivers_storage.go#L40-L125) for the attachment bits. Notice that `VolAttReqTrue` is really an alias for `VolAttReqWithDevMapOnlyVolsAttachedToInstance` which is `1 | 2 | 4 | 8` and requests attachment information before returning only volumes attached to the instance provided via the InstanceID header with an attempt to match the attachments to the provided local device information via the local devices header.